backend/cfg/eliminate_useless_edges.ml

```diff
@@ -0,0 +1,83 @@
+ * copies or substantial portions of the Software.                                *
+ *                                                                                *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR     *
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,       *
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE    *
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER         *
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,  *
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE  *
+ * SOFTWARE.                                                                      *
+ *                                                                                *
+ **********************************************************************************)
+[@@@ocaml.warning "+a-30-40-41-42"]
+
+(* Two blocks `b1` and `b2` can be merged if:
+   - `b1` has only one non-exceptional successor, `b2`;
+   - `b2` has only one predecessor, `b1`;
+   - `b1` and `b2` are distinct blocks.
+
+   When the condition is met, `b1` is modified as follows:
+   - its body is set to the concatenation of `b1.body` and `b2.body`;
+   - its terminator becomes the terminator of `b2`;
+   - its `exns`, `can_raise`, `can_raise_interproc`, `can_raise_interproc` fields
+     are set to the "union" of the respective fields in `b1` and `b2`;
+   - (its other fields are left unchanged);
+   and `b2` is modified as follows:
+   - its prececessors are set to empty;
+   - (its other fields are left unchanged).
+
+   As a consequence, `b2` becomes dead. *)
+
+let rec merge_blocks (modified : bool) (cfg : Cfg.t) =
+  let merged =
+    Label.Tbl.fold
+      (fun b1_label (b1_block : Cfg.basic_block) merged ->
+         let b1_successors = Cfg.successor_labels ~normal:true ~exn:false b1_block in
+         match Label.Set.cardinal b1_successors with
+         | 1 ->
+           let b2_label = Label.Set.choose b1_successors in
+           let b2_block = Label.Tbl.find cfg.blocks b2_label in
+           let b2_predecessors = Cfg.predecessor_labels b2_block in
+           if not (Label.equal b1_label b2_label)
+           && (List.compare_length_with b2_predecessors 1 = 0) then begin
+             assert (Label.equal b1_label (List.hd b2_predecessors));
+             (* modify b1 *)
+             b1_block.body <- b1_block.body @ b2_block.body;
+             b1_block.terminator <- b2_block.terminator;
+             b1_block.exns <- Label.Set.union b1_block.exns b2_block.exns;
+             b1_block.can_raise <- b1_block.can_raise || b2_block.can_raise;
+             b1_block.can_raise_interproc <- b1_block.can_raise_interproc || b2_block.can_raise_interproc;
+             (* modify b2 *)
+             b2_block.predecessors <- Label.Set.empty;
+             true
```

Comment:
I think you need to do `Disconnect_block.disconnect cfg_with_layout b2_label` here before return. 
For example, we should be able to merge  [b1]->[b2]->[b3]  into one block, but after "merge b1 b2" we can't merge b3 because it has 2 predecessors: the newly merged node and the old b2 hanging on.

Reply:
Indeed, that was embarrassing...